dali/test/python/test_pipeline_rnnt.py

```diff
@@ -0,0 +1,161 @@
+
+        self.log_features = ops.ToDecibels(device="cpu", multiplier=np.log(10), reference=1.0,
+                                           cutoff_db=-80)
+
+        self.get_shape = ops.Shapes(device="cpu")
+
+        self.normalize = ops.Normalize(axes=[0], device="cpu")
+
+        self.splicing_transpose = ops.Transpose(device="cpu", perm=[1, 0])
+        self.splicing_reshape = ops.Reshape(device="cpu", rel_shape=[-1, frame_splicing_factor])
+        self.splicing_pad = ops.Pad(axes=[0], fill_value=0, align=frame_splicing_factor, shape=[1],
+                                    device="cpu")
+
+        self.get_nonsilent_region = ops.NonsilentRegion(device="cpu", cutoff_db=silence_threshold)
+        self.trim_silence = ops.Slice(device="cpu", normalized_anchor=False, normalized_shape=False,
+                                      axes=[0], image_type=types.ANY_DATA)
+        self.to_float = ops.Cast(dtype=types.FLOAT)
+
+    @staticmethod
+    def _div_ceil(dividend, divisor):
+        return (dividend + divisor - 1) // divisor
+
+    def _splice_frames(self, input):
+        """
+        Frame splicing is implemented by transposing the input, padding it and reshaping.
+        Theoretically, to achieve the result now there should be one more transpose at the end,
+        but it can be skipped as an optimization
+        """
+        out = self.splicing_transpose(input)
+
+        # Because of the padding, we need to determine length of audio sample before it occurs
+        audio_len = self._div_ceil(self.get_shape(out), self.frame_splicing_factor)
+
+        out = self.splicing_pad(out)
+        out = self.splicing_reshape(out)
+        # Skipping transposing back
+        return out, audio_len
+
+    def remove_silence(self, input):
+        begin, len = self.get_nonsilent_region(input)
+        out = self.trim_silence(input, self.to_float(begin), self.to_float(len))
+        return out
+
+    def define_graph(self):
+        input, label = self.read()
+        # No resampling (aka "speed perturbation"), because of randomness of this operation
+        audio, sr = self.decode(input)
+
+        audio = self.remove_silence(audio)
+
+        audio = audio + self.normal_distribution(audio) * self.dither
+
+        # DALI's preemph works a little bit different than the one in native code.
+        # The difference occurs in first value in buffer.
+        audio = self.preemph(audio)
+
+        audio = self.spectrogram(audio)
+        audio = self.mel_fbank(audio)
+        audio = self.log_features(audio)
+        audio, audio_sh = self._splice_frames(audio)
+
+        # This normalization goes across ax=0, since
+        # the frame splicing doesn't transpose the tensor back
+        audio = self.normalize(audio)
+
+        return audio.gpu(), audio_sh.gpu()
+
+
+def test_rnnt_data_pipeline():
+    """
+    Test compares pre-calculated output of native data pipeline with an output
```

Comment:
Can we have this reference data generated online? How to generate this gold data on our own?

Reply:
It's kinda hard. We'd need code for RNNT model inside DALI, that should be modified for extracting intermediate buffers with preprocessed data. Plus, if we want actual data (not synthetic one), we'd need to tweak it even more (no easy way to give the same file in both data sources). Pre-generated reference was the easiest thing to do

Reply:
So as I understand it is evenly hard to regenerate this data. So if anything changes we cannot do that easily. I think it is far from perfect. Can't you just express the same transformation pipeline using librosa?

Reply:
Done